x/blob/test/decode_blob_tx_test.go

```diff
@@ -0,0 +1,111 @@
+package test
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/celestiaorg/celestia-app/app"
+	"github.com/celestiaorg/celestia-app/app/encoding"
+	"github.com/celestiaorg/celestia-app/pkg/blob"
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/assert"
+)
+
+type BlockResponse struct {
+	Block Block `json:"block"`
+}
+
+type Block struct {
+	Data BlockData `json:"data"`
+}
+
+type BlockData struct {
+	Txs []string `json:"txs"`
+}
+
+// TestDecodeBlobTx demonstrates how one can take the response from the Comet
+// BFT API endpoint /cosmos/base/tendermint/v1beta1/blocks/{block_number} and
+// convert a base64 encoded transaction into an sdk.Tx.
+//
+// NOTE: this process differs from other Cosmos SDK chains because the
+// transactions of type BlobTx won't be usable directly. One needs to extract
+// the Tx field inside the BlobTx prior to decoding it as an sdk.Tx.
+func TestDecodeBlobTx(t *testing.T) {
+	blockResponse := getBlockResponse(t)
+
+	for i, rawTx := range blockResponse.Block.Data.Txs {
+		decoded, err := base64Decode(rawTx)
+		if err != nil {
+			t.Errorf("error decoding base64 tx: %v", err)
+		}
+
+		txBytes := getTxBytes(decoded)
+		tx, err := decodeTx(txBytes)
+		if err != nil {
+			t.Errorf("error decoding tx: %v", err)
+		}
+
+		if i == 273 {
+			// https://celenium.io/tx/C55BDD3DF3348A9F8D9206528051804754F009A1B9D0F69CCC2F9D4334215D21
+			wantHash := "C55BDD3DF3348A9F8D9206528051804754F009A1B9D0F69CCC2F9D4334215D21"
+			gotHash := strings.ToUpper(hex.EncodeToString(sum(txBytes)))
+			assert.Equal(t, wantHash, gotHash)
+
+			wantSigner := "celestia18y3ydyn7uslhuxu4lcm2x83gkdhrrcyaqvg6gk"
+			gotSigner := tx.GetMsgs()[0].GetSigners()[0].String()
+			assert.Equal(t, gotSigner, wantSigner)
+		}
+	}
+}
+
+func getBlockResponse(t *testing.T) BlockResponse {
+	// block_response.json is the JSON response from the API endpoint:
+	// https://api.celestia.pops.one/cosmos/base/tendermint/v1beta1/blocks/408
+	// The response was persisted to block_response.json so that this test
+	// doesn't execute an HTTP request every invocation.
+	path := filepath.Join("testdata", "block_response.json")
+	fileContents, err := os.ReadFile(path)
+	if err != nil {
+		t.Fatalf("reading json file: %v", err)
+	}
+
+	var blockResponse BlockResponse
+	if err := json.Unmarshal(fileContents, &blockResponse); err != nil {
+		t.Fatalf("error decoding JSON block response: %v", err)
+	}
+	return blockResponse
+}
+
+func base64Decode(s string) ([]byte, error) {
+	return base64.StdEncoding.DecodeString(s)
+}
+
+func getTxBytes(txBytes []byte) []byte {
+	bTx, isBlob := blob.UnmarshalBlobTx(txBytes)
+	if isBlob {
+		return bTx.Tx
+	}
+	return txBytes
+}
+
+func decodeTx(txBytes []byte) (types.Tx, error) {
```

Comment:
[Question] Why not putting these utility functions in a main package instead of a test package?

Reply:
hmm because they're not used outside this test file

Reply:
Wouldn't they be useful?

Reply:
I guess they may be useful. After moving them to a non-test file, are you proposing also exporting them? If we don't export them then they won't be useful to external consumers. If we do export them, then we need to consider whether we should add to the exported public API.

I'm inclined to merge this as-is and create a follow-up issue if it's something we want to pursue.

Reply:
Yes, I meant also exporting them as they are likely to be used by the users if there is no such function available in the app.

Reply:
A follow up PR sounds good to me.